### PR TITLE
misc: drive-by code cleanup in BaseNode

### PR DESCRIPTION
**Pull request by brendankenny**
While starting work on robust TTI, there were some things that were a little hard to follow as I was stepping through graphs. Tried to make them a bit clearer in code and comments while keeping everything working the same.

@patrickhulce let me know if this feels like too much change for change's sake and I can scale it back :)

### CODE REVIEW

**Comment by brendankenny** [comment]
slightly better diff with [?w=1](https://github.com/GoogleChrome/lighthouse/pull/7723/files?w=1)

File: lighthouse-core/lib/dependency-graph/base-node.js [hunk 1]
```diff
@@ -168,96 +168,79 @@ class BaseNode {
...
 
-    /** @type {function(Node): boolean} */
-    let shouldIncludeNode = () => true;
-    if (predicate) {
-      const idsToInclude = new Set();
-      rootNode.traverse(node => {
-        if (predicate(node)) {
-          node.traverse(
-            node => idsToInclude.add(node.id),
-            node => node._dependencies.filter(parent => !idsToInclude.has(parent))
```
Comment (brendankenny) [inline lighthouse-core/lib/dependency-graph/base-node.js]: dangers of `Set<any>`...we were checking the `parent` node object against a set of string IDs, so weren't saving any time here

Comment (patrickhulce) [reply lighthouse-core/lib/dependency-graph/base-node.js]: nice fix!!

Comment (patrickhulce) [reply lighthouse-core/lib/dependency-graph/base-node.js]: I guess this one is difficult to test since it's just a marginal time savings huh

File: lighthouse-core/lib/dependency-graph/base-node.js [hunk 2]
```diff
@@ -168,96 +168,79 @@ class BaseNode {
...
+    // Walk down dependents.
+    rootNode.traverse(node => {
+      // Find new and includable nodes.
+      if (idsToIncludedClones.has(node.id)) return;
+      if (predicate && !predicate(node)) return;
+
+      // Walk back up dependencies, cloning all nodes in the path to this one.
+      node.traverse(
+        node => idsToIncludedClones.set(node.id, node.cloneWithoutRelationships()),
+        node => node._dependencies.filter(parent => !idsToIncludedClones.has(parent.id))
```
Comment (patrickhulce) [inline lighthouse-core/lib/dependency-graph/base-node.js]: can we get a comment on this one that explains the filter while we're at it? :) it even took me a sec to remember why this was happening at all

File: lighthouse-core/lib/dependency-graph/base-node.js [hunk 3]
```diff
@@ -168,96 +168,79 @@ class BaseNode {
...
-      if (!shouldIncludeNode(originalNode)) return;
-      const clonedNode = originalNode.cloneWithoutRelationships();
-      idToNodeMap.set(clonedNode.id, clonedNode);
+    // Walk down dependents.
+    rootNode.traverse(node => {
+      // Find new and includable nodes.
+      if (idsToIncludedClones.has(node.id)) return;
+      if (predicate && !predicate(node)) return;
+
+      // Walk back up dependencies, cloning all nodes in the path to this one.
```
Comment (patrickhulce) [inline lighthouse-core/lib/dependency-graph/base-node.js]: can we only do the reverse traversal when there's a predicate? 

that's one nice feature of the previous impl I wouldn't want to lose

File: lighthouse-core/lib/dependency-graph/base-node.js [hunk 4]
```diff
@@ -168,96 +168,79 @@ class BaseNode {
...
-   * visit with the getNext function.
-   * @param {function(Node, Node[]): void} iterator
-   * @param {function(Node): Node[]} [getNext] Defaults to returning the dependents.
-   */
-  traverse(iterator, getNext) {
-    if (!getNext) {
-      getNext = node => node.getDependents();
-    }
+    const visited = new Set([this.id]);
+    const queue = [[/** @type {Node} */(/** @type {BaseNode} */(this))]];
```
Comment (patrickhulce) [inline lighthouse-core/lib/dependency-graph/base-node.js]: do we still need this? I'm surprised we need to cast `BaseNode` for `this` hopefully this is something that was fixed :)

File: lighthouse-core/lib/dependency-graph/base-node.js [hunk 5]
```diff
@@ -168,96 +168,79 @@ class BaseNode {
...
+   * on each. `traversalPath` is the shortest (though not necessarily unique)
+   * path from `node` to the root of the iteration.
+   *
+   * The `getNextNodes` function takes a visited node and returns which nodes to
+   * visit next. It defaults to returning the node's dependents.
+   * @param {(node: Node, traversalPath: Node[]) => void} callback
+   * @param {function(Node): Node[]} [getNextNodes]
    */
-  _traversePaths(iterator, getNext) {
-    const stack = [[/** @type {Node} */(/** @type {BaseNode} */(this))]];
```
Comment (patrickhulce) [inline lighthouse-core/lib/dependency-graph/base-node.js]: lol I guess this is a queue :)

File: lighthouse-core/lib/dependency-graph/base-node.js [hunk 6]
```diff
@@ -168,96 +168,79 @@ class BaseNode {
...
 
-    getNext = node => {
-      visited.add(node.id);
-      const allNodesToVisit = originalGetNext(node);
-      const nodesToVisit = allNodesToVisit.filter(nextNode => !visited.has(nextNode.id));
-      nodesToVisit.forEach(nextNode => visited.add(nextNode.id));
-      return nodesToVisit;
-    };
+      for (const nextNode of getNextNodes(node)) {
+        if (visited.has(nextNode.id)) continue;
```
Comment (patrickhulce) [inline lighthouse-core/lib/dependency-graph/base-node.js]: having trouble finding the substantial changes here, it's just inlining the `getNext` rewrite from before and eliminating `traverse` right?

File: lighthouse-core/lib/dependency-graph/base-node.js (hunk 1 above)
Comment (brendankenny) [reply lighthouse-core/lib/dependency-graph/base-node.js]: > I guess this one is difficult to test since it's just a marginal time savings huh

yeah, and even at its worst it's really not that bad compared to other things we do with the graph

File: lighthouse-core/lib/dependency-graph/base-node.js (hunk 2 above)
Comment (brendankenny) [reply lighthouse-core/lib/dependency-graph/base-node.js]: > can we get a comment on this one that explains the filter while we're at it? 

good call

File: lighthouse-core/lib/dependency-graph/base-node.js (hunk 4 above)
Comment (brendankenny) [reply lighthouse-core/lib/dependency-graph/base-node.js]: > do we still need this? 

Annoyingly, yes. `this` *extends* `BaseNode`, but the lie is that it's a `Node`. Humans know the only types of `BaseNode` are `CPUNode | NetworkNode` so the cast to `Node` is ok, but the compiler accounts for the fact that we could be calling on a different class derived from `BaseNode`, so it's not necessarily compatible with `Node`.

So to cast while following the type assertion rules, we have to upcast to `BaseNode` and then downcast to `Node`. The other options is an assertion and a `@ts-ignore`...maybe that's more honest (albeit also ugly).

Comment (brendankenny) [reply lighthouse-core/lib/dependency-graph/base-node.js]: > Humans know the only types of `BaseNode` are `CPUNode | NetworkNode` so

I mean, even that's not entirely true. In tests we call `traverse()` with a graph of `BaseNode` nodes, but in real code it's much more convenient to only get `CPUNode | NetworkNode` back, so we should continue to pretend it's true :)

File: lighthouse-core/lib/dependency-graph/base-node.js (hunk 5 above)
Comment (brendankenny) [reply lighthouse-core/lib/dependency-graph/base-node.js]: > lol I guess this is a queue :)

your comment in `hasCycle` (which does DFS) underlined the humor: 

> // Get the last node in the stack (DFS uses stack, not queue)

I was thinking the approach changed at some point during the writing of `traverse` :)

File: lighthouse-core/lib/dependency-graph/base-node.js (hunk 6 above)
Comment (brendankenny) [reply lighthouse-core/lib/dependency-graph/base-node.js]: > it's just inlining the `getNext` rewrite from before and eliminating `_traversePaths` right?

yeah, main thing was the `getNext` rewriting and call to `_traversePaths` was breaking my brain when I was first (re)reading it, and it seemed like `traverse` has proven its design, so it seemed ok to inline at this point.

This was probably the bigger one for me saying it's ok to push back on changes if you don't think it reads better :)

*technically* we save n-1 `visited.add(node.id)` calls, but that's not worth much and could have been done without all the other things.

File: lighthouse-core/lib/dependency-graph/base-node.js (hunk 3 above)
Comment (brendankenny) [reply lighthouse-core/lib/dependency-graph/base-node.js]: > can we only do the reverse traversal when there's a predicate?

do you mean something like

```js
// Walk down dependents.
rootNode.traverse(node => {
  if (idsToIncludedClones.has(node.id)) return;

  if (!predicate) {
    // Clone every node.
    idsToIncludedClones.set(node.id, node.cloneWithoutRelationships());
  } else if (predicate(node)) {
    // Walk back up dependencies, cloning all nodes in the path to this one.
    node.traverse(
      node => idsToIncludedClones.set(node.id, node.cloneWithoutRelationships()),
      // Dependencies already cloned have already cloned ancestors, so no need to visit again.
      node => node._dependencies.filter(parent => !idsToIncludedClones.has(parent.id))
    );
  }
});
```

Or other ideas for formatting to make it less dense looking?

Another option is to make a default always true predicate, so it's a little clearer.

Agreed it is a bit weird to do the reverse traversal in the no predicate case. My thinking before was that we only have a single call of `cloneWithRelationships` without a predicate, so it didn't seem to be a super important case to call out (and since the filter line makes there be no next nodes for every node, it doesn't do any significant extra work)

File: lighthouse-core/lib/dependency-graph/base-node.js (hunk 5 above)
Comment (patrickhulce) [reply lighthouse-core/lib/dependency-graph/base-node.js]: haha yeah that was my guess as well when I was looking at this, I'll take your benefit of the doubt on this one lol 🤞

File: lighthouse-core/lib/dependency-graph/base-node.js (hunk 3 above)
Comment (patrickhulce) [reply lighthouse-core/lib/dependency-graph/base-node.js]: that's exactly what I meant :)

we could do an early return in the `!predicate` case to avoid the nesting? but IMO this is fine

File: lighthouse-core/lib/dependency-graph/base-node.js (hunk 6 above)
Comment (patrickhulce) [reply lighthouse-core/lib/dependency-graph/base-node.js]: nah, I like this. originally I had used `_traversePaths` in another method but that need hasn't materialized in the 1.5 years since so I think we're pretty safe :)

File: lighthouse-core/lib/dependency-graph/base-node.js (hunk 3 above)
Comment (brendankenny) [reply lighthouse-core/lib/dependency-graph/base-node.js]: done
